test/cpp/end2end/test_service_impl.cc

```diff
@@ -476,6 +479,9 @@ experimental::ServerUnaryReactor* CallbackTestServiceImpl::Echo(
       if (rpc_wait_thread_.joinable()) {
         rpc_wait_thread_.join();
       }
+      if (finish_when_cancelled_.joinable()) {
```

Comment:
Do you think this version of your change makes async_cancel_check_ redundant? Your change makes it so that we call ctx_->IsCancelled() asynchronously (but only in cancellation tests), so maybe that's sufficient for checking  if there are any data races associated with calling IsCancelled asynchronously.

Reply:
I'd like to keep that. As you pointed out, we're only checking ctx->IsCancelled in the Rpc function for actual cancellation cases whereas the AsyncCancelCheck thread checks it for both cancellation and non-cancellation cases. I can understand the merit of killing the async cancel check since technically its join is a potentially blocking operation (unlike the joins of finish when canceled and rpc wait thread) but I think it still provides test signal.